Problems cloning repository submodules

Update: the issue is to have the full repository clone, with submodules, work without submodule fatal errors. @gpx1000 volunteered to look at this, although they are not running into this on their own home network which may make reproduction difficult. However, both @SaschaWillems and myself are running into it frequently. Desired outcome is first, change the repository so this does not happen; or if that's not possible, give better advice in the README as to how to recover when it *does* happen.

I spun off the issue of how to build just the Vulkan-Samples component of docs.vulkan.org as #1181 as that's orthogonal to the submodule cloning problem.

----


When I try

> git clone --recurse-submodules git@github.com:KhronosGroup/Vulkan-Samples.git

as described in the README, I'm getting errors of the following form in the submodule cloning:

```
Cloning into 'Vulkan-Samples'...
Submodule 'assets' (https://github.com/KhronosGroup/Vulkan-Samples-Assets) registered for path 'assets'
...
Submodule 'third_party/vulkan' (https://github.com/KhronosGroup/Vulkan-Headers) registered for path 'third_party/vulkan'
...
Cloning into '/home/tree/git/Vulkan-Site/Vulkan-Samples/third_party/fmt'...
error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 7908 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/fmtlib/fmt' into submodule path '/home/tree/git/Vulkan-Site/Vulkan-Samples/third_party/fmt' failed
Failed to clone 'third_party/fmt'. Retry scheduled
Cloning into '/home/tree/git/Vulkan-Site/Vulkan-Samples/third_party/glfw'...
error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 7539 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/glfw/glfw' into submodule path '/home/tree/git/Vulkan-Site/Vulkan-Samples/third_party/glfw' failed
Failed to clone 'third_party/glfw'. Retry scheduled
Cloning into '/home/tree/git/Vulkan-Site/Vulkan-Samples/third_party/glm'...
...
Failed to clone 'third_party/imgui' a second time, aborting
```

I'll grant this could just be something about my ISP (AT&T Fiber, haven't had any issues in the last 10 months that weren't just temporary outages). But this operation feels very fragile - are there options that can help? If I then go into the clone and

> git submodule update

Then this does not complain. But the submodules that failed seem to just contain a .git directory and nothing else afterwards. Then

> cmake -H"." -B"build/unix" -DVKB_GENERATE_ANTORA_SITE=ON

gets a small distance into the build and starts failing with

```
...
-- Plugin `window_options` - BUILD
-- Configuring done
CMake Error at framework/CMakeLists.txt:462 (add_library):
  Cannot find source file:

    /home/tree/git/Vulkan-Site/Vulkan-Samples/third_party/CTPL/ctpl_stl.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
```

Here the CTPL directory just contains .git and nothing I've tried with submodule changes that situation. I can pull the underlying submodule's repository OK independently of the samples repo submodule setup - but not convince git that the submodule itself needs to be updated / replaced.

It would be really helpful if the README described these sorts of scenarios and advised how to work around them as this is far outside my minimal knowledge of submodules. I've tried removing Vulkan-Samples and re-cloning - the same sorts of failures (but not the same specific submodules) keep happening. It takes 20-30 minutes for a complete cycle with all the retries to complete and it's not easy to test.

--- a/README.adoc
+++ b/README.adoc
@@ -88,6 +88,21 @@ cd Vulkan-Samples
 
 Follow build instructions for your platform below.
 
+[NOTE]
+====
+The full repository is very large, and some ISPs appear to have trouble
+providing a robust connection to github while the clone is being made.
+
+If you notice problems such as submodules downloading at reported rates in
+the tens of kB/s, or fatal timeout errors occurring, these may be due to
+network routing issues to github within your ISP's internal network, rather
+than anything wrong in your own networking setup.
+
+It can be very difficult to get ISPs to acknowledge such problems exist much
+less to fix them. In this case, consider trying the clone while running a
+VPN service.
+====
+
 == Build
 
 === Supported Platforms

